ta: pkcs11: Update attributes in persistent storage for C_SetAttributeValue()

ruchi393 opened the pull request:
For token objects, for any modification in attributes,
the attributes also need to be updated in the objects
persistent storage. These modifications are done when
C_SetAttributeValue() is used.

Signed-off-by: Ruchika Gupta <ruchika.gupta@linaro.org>

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

vesajaaskelainen commented on ta/pkcs11/src/persistent_token.c:
```diff
@@ -538,6 +539,44 @@ void release_persistent_object_attributes(struct pkcs11_object *obj)
+		return tee2pkcs_error(res);
+	}
+
+	res = TEE_TruncateObjectData(hdl, size);
```
I have been thinking about the power cut endurance of the secure storage. Never really gone in the field deep enough yet but this got me thinking this once again.

I believe it would be better to:
- Open file
- Write data
- Truncate
- Close

This way file will always have correct data at the beginning (assuming that write at least goes to the storage) even if the truncate is lost.

Now if the truncate is made first and that succeeds and then power is lost then we have object that is possibly incomplete as new data is not yet written.

If power cut is in within the write operation then we have lost the game and that object is broken which is bad.

In order to tackle that I suppose we would need to create a new persistent object and then change that in place of old object and then delete the old object.

etienne-lms replied:
You got something here.
I think open/write/truncate/close looks the right sequence. And a sanity check/clean at database open may be needed to ensure perssitent object are clean. There would also address case where we create a new persistent object, got the object written in persistent object but its reference not written in the persistent dabase object because of some power loss, reboot or like.
I think you should address these case in a specific P-R.

ruchi393 replied:
Sequence you mentioned seems logical @vesajaaskelainen. I will correct it for this PR. But we have other instances in code where same sequence is followed. As @etienne-lms mentioned, we can address that as a separate PR.

vesajaaskelainen commented on ta/pkcs11/src/object.c:
```diff
@@ -999,9 +999,14 @@ enum pkcs11_rc entry_set_attribute_value(struct pkcs11_client *client,
 	if (rc)
 		goto out;
 
+	if (get_bool(obj->attributes, PKCS11_CKA_TOKEN)) {
```
Good catch.

etienne-lms replied:
indeed! :)

vesajaaskelainen commented on ta/pkcs11/src/persistent_token.c:
```diff
@@ -538,6 +539,44 @@ void release_persistent_object_attributes(struct pkcs11_object *obj)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_ObjectHandle hdl = TEE_HANDLE_NULL;
+	uint32_t tee_obj_flags = TEE_DATA_FLAG_ACCESS_READ |
```
Do we need read in here?

etienne-lms replied:
You're right. For Write/Truncate, we don't need ACCESS_READ.

ruchi393 replied:
Thanks for noticing it. I will remove both ACCESS_READ and TEE_DATA_FLAG_ACCESS_WRITE_META

etienne-lms reviewed (commented):
oups... I had some pending review comments I forgot to post. Here there are.

ruchi393 commented on ta/pkcs11/src/persistent_token.c:
```diff
@@ -538,6 +539,44 @@ void release_persistent_object_attributes(struct pkcs11_object *obj)
+				 TEE_DATA_FLAG_ACCESS_WRITE_META;
+	size_t size = 0;
+
+	if (obj->attributes)
```
My bad. Thanks @etienne-lms . I will correct this.

etienne-lms commented on ta/pkcs11/src/persistent_token.c:
```diff
@@ -573,6 +566,9 @@ enum pkcs11_rc update_persistent_object_attributes(struct pkcs11_object *obj)
 
+	res = TEE_TruncateObjectData(hdl, size);
+	if (res)
+		goto out;
```
can drop these 2 lines

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` with minor comment addressed or not.

ruchi393 commented:
Thanks @etienne-lms .

With addition of assert, gcc gives error due to following warning :
ta/pkcs11/src/persistent_token.c:542:16: error: function might be candidate for attribute 'noreturn' [-Werror=suggest-attribute=noreturn]
 enum pkcs11_rc update_persistent_object_attributes(struct pkcs11_object *obj)
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

I will need to change it to this to avoid this warning
 
-       assert(!obj->attributes);
+       if (!obj->attributes)
+               assert(!obj->attributes);
 
Would this be acceptable ?

etienne-lms commented on ta/pkcs11/src/persistent_token.c:
```diff
@@ -538,6 +539,40 @@ void release_persistent_object_attributes(struct pkcs11_object *obj)
+	size_t size = 0;
+
+	if (!obj->attributes)
+		assert(!obj->attributes);
```
ok, sorry. To catch gcc warning, I think `assert(obj && obj->attributes);` is the right implementation :), thank you gcc!

and maybe move `size = sizeof(struct obj_attrs) + obj->attributes->attrs_size;` to line 566, before it's used.

ruchi393 replied:
My bad. Sorry. I don't know what i was thinking to add a negation with obj->attributes in assert. I will correct that.

etienne-lms commented on ta/pkcs11/src/persistent_token.c:
```diff
@@ -538,6 +539,40 @@ void release_persistent_object_attributes(struct pkcs11_object *obj)
+out:
+	TEE_CloseObject(hdl);
+	return tee2pkcs_error(res);
+}
```
add an empty line below

etienne-lms replied:
you forgot this one :)

ruchi393 replied:
Sorry again. I misunderstood that you wanted a blank line between 573 and 574 and added that. I am removing that now and adding one after 575.

etienne-lms reviewed (changes requested):
some stuff to fix

ruchi393 commented:
Fixed issues and addressed review comments

etienne-lms commented on ta/pkcs11/src/object.c:
```diff
@@ -999,9 +999,14 @@ enum pkcs11_rc entry_set_attribute_value(struct pkcs11_client *client,
+
 	DMSG("PKCS11 session %"PRIu32": set attributes %#"PRIx32,
 	     session->handle, object_handle);
-
```
removing this line is not needed. I prefer you discard this suppression so that commit only does what it pretends to: add a missing piece.

etienne-lms commented:
2 remaining and I'm done.

ruchi393 commented:
Fixed the last 2 issues, squashed and resubmitted the PR.

@etienne-lms , thanks for reviewing. You had given a R-by earlier. Should I add that now ?

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

ruchi393 commented:
Thanks @etienne-lms . Tag applied.

jforissier commented:
Clearly unrelated to this PR -- but it it the second time at least I see this "segmentation fault" message in IBART when running xtest 1033 (`tee-supplicant` plugin test).
https://optee.mooo.com:5000/logs/OP-TEE/optee_os/4403/575644055/c7bb97539a08b7e18e4cdd34904fc5e9f6af7342#end-01-hikey6220-test

@anisyanka any idea what might be wrong?

anisyanka commented:
> @anisyanka any idea what might be wrong?

All tests have been successfully passed in the link above. I don't see  "segmentation fault" message, but maybe the tests were restarted before I had a look and the subtle bug was already gone.

Which devices have you seen this malfunction on? Or it's a random process?
I want to try to catch the bug in qemu, but for now in qemu-v8 i've never seen the segfault.

jforissier commented:
@anisyanka 
> All tests have been successfully passed in the link above. I don't see "segmentation fault" message, but maybe the tests were restarted before I had a look and the subtle bug was already gone.

Oh, right, I suppose I have manually restarted the IBART run (I don't remember for sure) and it would have cleared the previous log. @jbech-linaro does IBART keep logs only as new commits are pushed in the same PR, but not when manually restarting the build on the same commit?

> Which devices have you seen this malfunction on? Or it's a random process?

I have seen the test fail a couple of times at least on HiKey6220 and it appears to be random and not so frequent. I'll try to dig up an old log.

jforissier commented:
> I have seen the test fail a couple of times at least on HiKey6220 and it appears to be random and not so frequent. I'll try to dig up an old log.

This one: https://optee.mooo.com:5000/logs/OP-TEE/optee_os/4410/577146745/ffcc5ec42c627edf555766a8c0963cd205d4e664#end-01-hikey6220-test

anisyanka commented:
> > I have seen the test fail a couple of times at least on HiKey6220 and it appears to be random and not so frequent. I'll try to dig up an old log.
> 
> This one: https://optee.mooo.com:5000/logs/OP-TEE/optee_os/4410/577146745/ffcc5ec42c627edf555766a8c0963cd205d4e664#end-01-hikey6220-test

Thanks, I'll try to catch and debug this

jforissier commented:
@vesajaaskelainen asking since you commented, do you want to provide a review tag?

vesajaaskelainen commented on ta/pkcs11/src/persistent_token.c:
```diff
@@ -538,6 +539,40 @@ void release_persistent_object_attributes(struct pkcs11_object *obj)
+		return tee2pkcs_error(res);
+	}
+
+	res = TEE_SeekObjectData(hdl, 0, TEE_DATA_SEEK_SET);
```
In the specification for TEE_OpenPersistentObject:
`The initial data position in the data stream is set to 0.`

So we don't need to do the seek to start.

ruchi393 replied:
Thanks, you are right. I will remove the Seek call. @etienne-lms , is that ok with you also ?

etienne-lms replied:
yes, agree.

vesajaaskelainen commented:
> @vesajaaskelainen asking since you commented, do you want to provide a review tag?

Added one comment about unnecessary TEE_SeekObjectData() call.

With that addressed otherwise looks good:
`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

ruchi393 commented:
Removed TEE_SeekObjectData() as suggested by @vesajaaskelainen .
Tags applied.
